pkg/docker/docker.go

```diff
@@ -608,70 +606,81 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) error {
-		return err
-	}
-
-	// Start the container.
-	glog.V(2).Infof("Starting container %q ...", containerName)
-	if err := util.TimeoutAfter(DefaultDockerTimeout, "timeout after waiting %v for Docker to start container", func() error {
-		return d.client.StartContainer(container.ID, nil)
-	}); err != nil {
-		return err
+		removalDone <- struct{}{}
 	}
+	dumpStack := func(signal os.Signal) {
+		if signal == syscall.SIGQUIT {
+			buf := make([]byte, 1<<16)
+			runtime.Stack(buf, true)
+			fmt.Printf("%s", buf)
+		}
+		os.Exit(1)
+	}
+	return util.NewInterruptHandler(dumpStack, removeContainer).Run(func() error {
+		// Attach to the container.
+		glog.V(2).Infof("Attaching to container %q ...", containerName)
+		attachOpts := opts.asDockerAttachToContainerOptions()
+		attachOpts.Container = container.ID
+		if _, err = d.client.AttachToContainerNonBlocking(attachOpts); err != nil {
+			glog.Errorf("Unable to attach to container %q with options %+v: %v", containerName, attachOpts, err)
+			return err
+		}
 
-	// Run OnStart hook if defined.
-	if opts.OnStart != nil {
-		if err = opts.OnStart(container.ID); err != nil {
+		// Start the container.
+		glog.V(2).Infof("Starting container %q ...", containerName)
+		if err := util.TimeoutAfter(DefaultDockerTimeout, "timeout after waiting %v for Docker to start container", func() error {
+			return d.client.StartContainer(container.ID, nil)
+		}); err != nil {
 			return err
 		}
-	}
 
-	// We either block waiting for a user-originated SIGINT, or wait for the
-	// container to terminate. When TargetImage is true, we're dealing with
-	// an invocation of `s2i build ... --run` so this will, e.g., run a web
-	// server and block until the user interrupts it. The other case is seen
-	// when running the assemble script or other commands that are meant to
-	// terminate in a finite amount of time.
-	if opts.TargetImage {
-		runContainerDockerRun(container, d, image)
-	} else {
-		// Return an error if the exit code of the container is
-		// non-zero.
-		glog.V(4).Infof("Waiting for container %q to stop ...", containerName)
-		exitCode, err := d.client.WaitContainer(container.ID)
-		if err != nil {
-			return fmt.Errorf("waiting for container %q to stop: %v", containerName, err)
+		// Run OnStart hook if defined.
+		if opts.OnStart != nil {
+			if err = opts.OnStart(container.ID); err != nil {
+				return err
+			}
 		}
-		if exitCode != 0 {
-			return errors.NewContainerError(container.Name, exitCode, "")
+
+		// We either block waiting for a user-originated signal, or wait for the
+		// container to terminate. When TargetImage is true, we're dealing with
+		// an invocation of `s2i build ... --run` so this will, e.g., run a web
+		// server and block until the user interrupts it. The other case is seen
+		// when running the assemble script or other commands that are meant to
+		// terminate in a finite amount of time.
+		if opts.TargetImage {
+			dumpContainerInfoAndWait(container, d, image)
```

Comment:
@bparees so the difference in behavior that I think we're introducing is that:

**BEFORE**

We handle a SIGINT which causes `dumpContainerInfoAndWait` (with its old name) to return. The program flow follows normally to the next line after this `if` block.

**AFTER**

With the changes we have so far, we will catch a SIGINT and call `os.Exit(1)`, meaning that line 683+ never sees the light of day... Currently this means `PostExec` never runs.


----------------

We need to check if there's any case today where `opts.TargetImage == true && opts.PostExec != nil`.

That would also be a rather tricky code for us to change in the future, as we'll have to have in mind that the only way for `dumpContainerInfoAndWait` to "return" is through a `os.Exit(0)`, meaning all code after it is unreachable.